File: leetcode-solutions/ruby/373. Find K Pairs with Smallest Sums.rb
# https://leetcode.com/problems/find-k-pairs-with-smallest-sums/

# TLE solution with minHeap
require 'algorithms'
include Containers
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @param {Integer} k
# @return {Integer[][]}
def k_smallest_pairs(nums1, nums2, k)
    min_heap = MinHeap.new
    [k, nums1.size].min.times do |i|
        min_heap.push([nums1[i] + nums2[0], nums1[i], nums2[0], 0])
    end
    ans = []
    k.times do
        break if min_heap.empty?
        _, n1, n2, idx = min_heap.pop
        ans.push([n1, n2])
        if idx + 1 < nums2.size
            min_heap.push([n1+nums2[idx+1], n1, nums2[idx+1], idx+1])
        end
    end
    
    ans
end

File: leetcode-solutions/python/373. Find K Pairs with Smallest Sums.py
# https://leetcode.com/problems/find-k-pairs-with-smallest-sums/

from heapq import heappop, heappush


class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        M = len(nums1)
        N = len(nums2)
        heap = []
        ans = []
        # first take all the element from nums1
        for i in range(M):
            self.add(heap, nums1, nums2, i, 0)

        while k > 0 and heap:
            _, i, j = heappop(heap)
            ans.append([nums1[i], nums2[j]])
            self.add(heap, nums1, nums2, i, j+1)
            k -= 1

        return ans

    def add(self, heap, nums1, nums2, i, j):
        if j >= len(nums2):
            return
        heappush(heap, (nums1[i] + nums2[j], i, j))
